src/agent.js

```diff
@@ -46,56 +51,185 @@ async function generateAllBots(_config) {
   return botConfigs;
 }
 
-async function initialize() {
-  const botConfigs = await generateAllBots(config);
+/*
+  gatherType "all" is weird. A quick rundown:
 
-  const botStateProms = botConfigs.map((bot) => {
-    const botMod = botMap.get(bot.botType);
-    if (botMod.initialize === undefined) {
-      const botState = { ...bot };
-      return new Promise(() => botState);
-    }
+  Inbound blocks look like this:
+  HandleBlock
+  - HandleTransaction
+  - HandleTransaction
+  - HandleTransaction
+  - HandleTransaction
 
-    return botMod.initialize(bot);
-  });
+  Notable outbound cases:
 
-  const results = await Promise.all(botStateProms);
-  results.forEach((result) => botStates.push(result));
-}
+  for all positive block findings, where we only have block handling bots:
+    HandleBlock -> [Finding, Finding, Finding]
+    - HandleTransaction -> []
 
-function handleAllTransactions(_botMap, _botStates) {
-  return async function handleTransaction(txEvent) {
-    const findProms = _botStates.map((bot) => {
+  for all positive tx findings, where we only have tx handling bots:
+    HandleBlock -> []
+    - HandleTransaction -> [Finding, Finding, Finding]
+
+  for all positive tx and block findings:
+    HandleBlock -> []
+    - HandleTransaction -> [...blockFindings[Finding], ...txFindings[Finding, Finding]]
+
+  for all positive block findings, some/no tx findings:
+    HandleBlock -> []
+    - HandleTransaction -> []
+
+  for only some positive block findings, all/some/no tx findings:
+    HandleBlock -> []
+    - HandleTransaction -> []
+
+  Bots can have a block handler, a transaction handler, or both
+
+  if there are only bots with block handlers:
+    - run the block handlers, aggregate the results
+    - if all bots return findings, return the findings
+    - else return []
+  if there are only bots with tx handlers:
+    - run the tx handlers, aggregate the results
+    - if all bots return findings, return the findings
+    - else return []
+
+  if you've got mixed tx and block handlers:
+  for all bots with block handlers:
+    - run the handlers, aggregate the results
+    - if all bots return findings, cache that for the tx handlers
+  for all bots with tx handlers:
+    - run the handlers, aggregate the results
+    - if there are no block handlers, and all bots have tx findings, return the findings
+    - if there are block handlers, and all bots have tx findings, return the findings
+*/
+
+function initializeBots(_config, _botMap, _botStates) {
+  return async function initialize() {
+    const botConfigs = await generateAllBots(_config, _botMap);
+
+    /* eslint-disable no-param-reassign */
+    _botStates.gatherMode = config.gatherMode;
```

Comment:
This looks like it should come from the passed in `_config` argument, rather than the global `config` variable.
```suggestion
    _botStates.gatherMode = _config.gatherMode;
```